docs/src/lens.md

```diff
@@ -0,0 +1,276 @@
+#[derive(Clone, Data, Lens)]
+struct TodoItem {
+    title: String,
+    completed: bool,
+    urgent: bool,
+}
+
+fn make_todo_item() -> impl Widget<TodoItem> {
+    // A label that generates its text based on the data
+    let title = Label::dynamic(|text, _| text.to_string()).lens(TodoItem::title);
+    let completed = Checkbox::new().lens(TodoItem::completed);
+    let urgent = Checkbox::new().lens(TodoItem::urgent);
+
+    Flex::column()
+        // label on top
+        .with_child(title)
+        // two checkboxes below
+        .with_child(Flex::row().with_child(completed).with_child(urgent))
+}
+```
+
+## Advanced lenses
+
+Field access is a very simple (and common, and *useful*) case, but lenses can do much more than that.
+
+### `LensExt` and combinators
+
+Similar to the `WidgetExt` trait, we offer a `LensExt` trait that provides
+various functions for composing lenses. These are similar to the various methods
+on iterator; you can `map` from one lens to another, you can index into a
+collection, or you can efficiently access data in an `Arc` without unnecessary
+mutation; see the main crate documentation for more.
+
+As your application gets more complicated, it will become likely that you want
+to use fancier sorts of lensing, and `map` and company can start to get out
+of hand; when that happens, you can always implement a lens by hand.
+
+### getting something from a collection
+
+Let say your application is a contact book, and you would like a lens that
+focuses on a specific contact. A simplified version of our data looks like this:
+
+```rust
+#[derive(Clone, Data)]
+struct Contact {
+    // fields
+}
+
+type ContactId = u64;
+
+#[derive(Clone, Data)]
+struct Contacts {
+    inner: Arc<HashMap<ContactId, Contact>>,
+}
+
+// Lets write a lens that returns a specific contact based on its id, if it exists.
+
+struct ContactIdLens(ContactId);
+
+impl Lens<Contacts, Option<Contact>> for ContactIdLens {
+    fn with<R, F: FnOnce(&Option<Contact>) -> R>(&self, data: &Contacts, f: F) -> R {
+        let contact = data.inner.get(&self.0).cloned();
+        f(&contact)
+    }
+
+    fn with_mut<R, F: FnOnce(&mut Option<Contact>) -> R>(&self, data: &mut Contacts, f: F) -> R {
+        // get an immutable copy
+        let mut contact = data.inner.get(&self.0).cloned();
+        let result = f(&mut contact);
+        // only actually mutate the collection if our result is mutated;
+        if !data.inner.get(&self.0).same(&contact.as_ref()) {
+            let contacts = Arc::make_mut(&mut data.inner);
+            // if we're none, we were deleted, and remove from the map; else replace
+            match contact {
+                Some(contact) => contacts.insert(self.0, contact),
+                None => {
+                    contacts.remove(self.0);
+                }
+            }
+        }
+        result
+    }
+}
+```
+
+### doing a conversion
+
+What if you have a distance in miles that you would like to display in
+kilometres?
```

Comment:
I'm assuming we're using US English, so this should be `kilometers`.

Reply:
As a canadian I vacillate between american & british spelling, and my personal position here is that if the US isn't going to *use* the metric system it has relinquished its authority over how metric units should be spelt. I mean spelled.